boards/cpu: add support for nucleo144-f746 and stm32f7

haukepetersen opened the pull request:
Almost there, but no UART output. Seems like I missed something, but can't see it right now. @aabadie: feel free to take over this PR - or wait until I am back from holiday...

DipSwitch commented:
Did you switch the solder jumper configuration?

On Apr 13, 2017 22:32, "Hauke Petersen" <notifications@github.com> wrote:

> Almost there, but no UART output. Seems like I missed something, but can't
> see it right now. @aabadie <https://github.com/aabadie>: feel free to
> take over this PR - or wait until I am back from holiday...
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/RIOT-OS/RIOT/pull/6907
> Commit Summary
>
>    - core/panic: enable panic modes for Cortex-M7
>    - cpu/cortex_common: added support for Cortex-M7
>    - cpu/stm32: added support for STM32F7
>    - boards: added support for nucleo144-f746
>
> File Changes
>
>    - *A* boards/nucleo144-f746/Makefile
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-0> (3)
>    - *A* boards/nucleo144-f746/Makefile.dep
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-1> (1)
>    - *A* boards/nucleo144-f746/Makefile.features
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-2> (12)
>    - *A* boards/nucleo144-f746/Makefile.include
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-3> (6)
>    - *A* boards/nucleo144-f746/board.c
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-4> (34)
>    - *A* boards/nucleo144-f746/dist/openocd.cfg
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-5> (5)
>    - *A* boards/nucleo144-f746/include/board.h
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-6> (35)
>    - *A* boards/nucleo144-f746/include/periph_conf.h
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-7> (147)
>    - *M* core/include/panic.h
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-8> (2)
>    - *M* cpu/cortexm_common/cortexm_init.c
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-9> (4)
>    - *M* cpu/cortexm_common/include/vectors_cortexm.h
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-10> (2)
>    - *M* cpu/cortexm_common/thread_arch.c
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-11> (6)
>    - *M* cpu/cortexm_common/vectors_cortexm.c
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-12> (2)
>    - *M* cpu/stm32_common/include/periph_cpu_common.h
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-13> (5)
>    - *M* cpu/stm32_common/periph/uart.c
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-14> (9)
>    - *A* cpu/stm32f7/Makefile
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-15> (10)
>    - *A* cpu/stm32f7/Makefile.include
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-16> (5)
>    - *A* cpu/stm32f7/cpu.c
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-17> (37)
>    - *A* cpu/stm32f7/include/cpu_conf.h
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-18> (47)
>    - *A* cpu/stm32f7/include/periph_cpu.h
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-19> (85)
>    - *A* cpu/stm32f7/include/stmclk.h
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-20> (82)
>    - *A* cpu/stm32f7/include/vendor/stm32f746xx.h
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-21> (17565)
>    - *A* cpu/stm32f7/ldscripts/stm32f746zg.ld
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-22> (30)
>    - *A* cpu/stm32f7/periph/Makefile
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-23> (5)
>    - *A* cpu/stm32f7/stmclk.c
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-24> (149)
>    - *A* cpu/stm32f7/vectors.c
>    <https://github.com/RIOT-OS/RIOT/pull/6907/files#diff-25> (322)
>
> Patch Links:
>
>    - https://github.com/RIOT-OS/RIOT/pull/6907.patch
>    - https://github.com/RIOT-OS/RIOT/pull/6907.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/6907>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAdFRJ9_YcIaGoSvYj6R1fHpOMO6GTtXks5rvoZegaJpZM4M9K4D>
> .
>

aabadie commented on cpu/stm32f7/ldscripts/stm32f746zg.ld:
```diff
@@ -0,0 +1,30 @@
+{
+    rom (rx)    : ORIGIN = 0x08000000, LENGTH = 1024K
+    ram (rwx)   : ORIGIN = 0x20000000, LENGTH = 320K
+    cpuid (r)   : ORIGIN = 0x1ff80050, LENGTH = 12
```
this should be `0x1ff0f420` see [datasheet](http://www.st.com/content/ccc/resource/technical/document/reference_manual/c5/cf/ef/52/c0/f1/4b/fa/DM00124865.pdf/files/DM00124865.pdf/jcr:content/translations/en.DM00124865.pdf) section 41.1 page 1654

aabadie commented on cpu/stm32f7/include/cpu_conf.h:
```diff
@@ -0,0 +1,47 @@
+}
+#endif
+
+#endif /* STM32L4_CPU_CONF_H */
```
F7

aabadie commented on cpu/stm32f7/vectors.c:
```diff
@@ -0,0 +1,322 @@
+ISR_VECTORS const void *interrupt_vector[] = {
+    /* Exception stack pointer */
+    (void*) (&_estack),             /* pointer to the top of the stack */
+    /* Cortex-M4 handlers */
```
M7

aabadie commented:
I allowed myself to push a couple of commits that fix a few problems (cpuid + common nucleo-144 board include). But still no UART or GPIO.

haukepetersen commented:
@aabadie: thanks for the fixes! Just out of curiosity: were you able to push directly into my branch without having write access? Interesting :-)

@DipSwitch: nope, did not touch the default hardware configuration, must I?

aabadie commented:
> Just out of curiosity: were you able to push directly into my branch without having write access? Interesting :-)

This is one of the [GitHub's features](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork). When one creates a PR, there's a checkbox for allowing maintainers of a project to push someone else's branch.

haukepetersen commented:
Ahh, I see, very interesting, good to know (and quite useful!) :-)

aabadie commented:
Yes but this is something that we should not abuse, especially with new contributors (even though this is debatable). Otherwise I think it is useful when we want to speed things up by just fixing remaining nits or squashing just before merging. But this is not a very interesting job.

vincent-d commented:
I have that board: http://www.st.com/en/evaluation-tools/32f746gdiscovery.html
So I tried to port it based on that PR. I'm also not able to have a working UART...

I found an issue with `periph_clk_en` and `periph_clk_dis` in `stm32_common/cpu_common.c`: AHB1 is not enabled for stm32f7. Fixing that should fix the GPIO, but it does not solve the UART issue for me.

haukepetersen commented:
Alright, after some more hours of debugging I got it to work! Should run fine now. There were actually a couple of issues: wrong flash wait cycle computation, PLL was not set as system clock, HSE was not properly enabled... But all fixed now.

haukepetersen commented:
Can anyone explain to me, where the `cppcheck` warning comes from? I thought we got rid of any `cppcheck` warnings in the code base, so that this (unrelated to this PR) warnings should not happen again?

And anyway, how do we get around this false positive?

miri64 commented:
Add a suppression comment

```C
/* cppcheck-suppress variableScope
 * variable used in assembly-code below */
```

miri64 commented:
(above the line pointed out by cppcheck).

Regarding the supposed fixing: did anyone test locally with the version of cppcheck running on Murdock?

aabadie commented:
> Alright, after some more hours of debugging I got it to work! Should run fine now. There were actually a couple of issues: wrong flash wait cycle computation, PLL was not set as system clock, HSE was not properly enabled... But all fixed now.

Good job ! I'll test asap

aabadie commented on cpu/stm32f7/include/periph_cpu.h:
```diff
@@ -0,0 +1,85 @@
+
+#ifndef DOXYGEN
+/**
+ * @brief   Override ADC resolution values
```
`@name`
btw, I'm wondering if those ADC definitions are required since this MCU doesn't provide the implementations for ADC ?

miri64 replied:
No, brief is right, but the braces are not required here (since doxygen parses the enum as one type anyway)

miri64 replied:
Then again, doesn't `#ifndef DOXYGEN` imply that this is NOT parsed by doxygen. In this case `@brief` still makes more sense, since this is a *brief* description, not a name.

haukepetersen replied:
removed them for now.

vincent-d commented on cpu/stm32_common/cpu_common.c:
```diff
@@ -60,7 +60,7 @@ void periph_clk_en(bus_t bus, uint32_t mask)
             break;
 #elif defined(CPU_FAM_STM32F2) || defined(CPU_FAM_STM32F4) \
-            || defined(CPU_FAM_STM32L4)
+            || defined(CPU_FAM_STM32L4) || defined(CPU_FAM_STM32F7)
```
Should be added in `periph_clk_dis` as well

haukepetersen replied:
fixed

aabadie replied:
indentation looks off

aabadie commented on cpu/stm32f7/vectors.c:
```diff
@@ -0,0 +1,322 @@
+ */
+
+/**
+ * @ingroup     cpu_stm32l4
```
cpu_stm32f7

aabadie commented on cpu/stm32f7/vectors.c:
```diff
@@ -0,0 +1,322 @@
+WEAK_DEFAULT void isr_svc(void);
+WEAK_DEFAULT void isr_pendsv(void);
+WEAK_DEFAULT void isr_systick(void);
+/* STM32L4 specific interrupt vectors */
```
STM32F7

aabadie commented on cpu/stm32f7/vectors.c:
```diff
@@ -0,0 +1,322 @@
+    (void*) isr_fpu
+};
+
+#if 0
```
I guess this block can be removed

aabadie commented on boards/nucleo144-f746/include/periph_conf.h:
```diff
@@ -0,0 +1,147 @@
+    {
+        .dev        = USART2,
+        .rcc_mask   = RCC_APB1ENR_USART2EN,
+        .rx_pin     = GPIO_PIN(PORT_A, 3),
```
I would prefer to use PG9 and PG14 for this UART. Then it corresponds to the Arduino RX/TX pins. See what's done the the other nucleo144 boards.

haukepetersen replied:
changed pins to PD5 and PD6

aabadie commented on boards/nucleo144-f746/include/periph_conf.h:
```diff
@@ -0,0 +1,147 @@
+    {
+        .dev        = USART1,
+        .rcc_mask   = RCC_APB2ENR_USART1EN,
+        .rx_pin     = GPIO_PIN(PORT_B, 7),
```
Same here there's a USART text printed next to CN8, I would prefer to use one the USART pins available there (probably PD6,PD5)

haukepetersen replied:
cant find any more useful mapping these pins...

aabadie replied:
use USART2, AF7 instead

aabadie commented on boards/nucleo144-f746/Makefile.features:
```diff
@@ -0,0 +1,12 @@
+FEATURES_PROVIDED += periph_cpuid
+FEATURES_PROVIDED += periph_gpio
+FEATURES_PROVIDED += periph_hwrng
+FEATURES_PROVIDED += periph_timer
```
For completeness, we could add some PWM configuration. Maybe it's better to add it in a follow-up PR ?

haukepetersen replied:
I would prefer doing it in a follow-up

aabadie replied:
ok

aabadie commented:
I also noticed the rework of clock initialization that I think is a nice addition. Do you think it could be generalized to all STM32 ?

aabadie commented:
Tested on board: no UART. Is there any special change required on the board or can I use the default configuration ?

haukepetersen commented:
> I also noticed the rework of clock initialization that I think is a nice addition. Do you think it could be generalized to all STM32 ?

Yes, thats the point of it. But lets go in small steps, the next one being in #6970

haukepetersen commented:
> Tested on board: no UART. Is there any special change required on the board or can I use the default configuration ?

Nope, didn't touch any (HW) configuration on the board. Mine works as is with this branch.

kYc0o commented:
@aabadie Try to update the STLink firmware, maybe that's the problem.

aabadie commented:
> Try to update the STLink firmware, maybe that's the problem.

Good idea, but doesn't work...

aabadie commented on boards/nucleo144-f746/include/periph_conf.h:
```diff
@@ -0,0 +1,147 @@
+#endif
+    },
+    {
+        .dev        = USART2,
```
Looking at the board pinout, if we want to be consistent with other nucleo144 (see #6944), the second UART configured should be USART6 on PG9 and PG14, AF8 (see table 12 in datasheet page 83 and 84).

aabadie commented on cpu/stm32_common/cpu_common.c:
```diff
@@ -108,7 +108,7 @@ void periph_clk_dis(bus_t bus, uint32_t mask)
             break;
 #elif defined(CPU_FAM_STM32F2) || defined(CPU_FAM_STM32F4) \
-            || defined(CPU_FAM_STM32L4)
+            || defined(CPU_FAM_STM32L4) || defined(CPU_FAM_STM32F7)
```
indentation looks off

aabadie commented on cpu/stm32f7/include/stmclk.h:
```diff
@@ -0,0 +1,82 @@
+#endif
+
+/**
+ * @brief   Configure the  high speed clock domain (main system clock)
```
there's an extra space between `the` and `high`

haukepetersen commented:
- adapted UART config to be in line with `nucleo144-f446`
- rebased on top of #6970 to make use of the shared `stmclk.h` header
- fixed indention in `cpu_common.c`

haukepetersen commented:
regarding the UART, @aabadie: did you try to connect an external UART-to-USB adapter to the rx/tx pins of UART3 (PD8/PD9) and try if you see something there?

aabadie commented:
> did you try to connect an external UART-to-USB adapter to the rx/tx pins of UART3 (PD8/PD9) and try if you see something there?

Good idea but I didn't try yet. I'll see if I have some time at the end of the afternoon.

aabadie commented on boards/nucleo144-f746/include/periph_conf.h:
```diff
@@ -0,0 +1,147 @@
+
+#define UART_0_ISR          (isr_usart3)
+#define UART_0_DMA_ISR      (isr_dma1_stream6)
+#define UART_1_ISR          (isr_usart2)
```
should be `isr_usart6` here and `isr_isart2` below

haukepetersen replied:
yes

haukepetersen commented:
rebased and fixed UART ISRs

aabadie commented:
@haukepetersen, I just gave another try with the latest version of this PR but still no UART or LED blinking. Just to be sure, I flashed a firmware built with mbed and it worked (well!).

Using OpenOCD, I noticed that the RIOT firmware doesn't reach the main function: it remains stuck in `cpu_switch_context_exit` called from kernet_init.

aabadie commented on cpu/stm32f7/ldscripts/stm32f746zg.ld:
```diff
@@ -0,0 +1,30 @@
+ */
+
+/**
+ * @addtogroup      cpu_stm32l4
```
cpu_stm32f7

haukepetersen commented:
good catch -> seems there was a reason I had that block of out-commented interrupt vectors still in the the vectors file earlier... Fixed.

Also explicitly tested UART_DEV(1) and confirmed it working:
```
2017-05-04 10:47:23,536 - INFO # main(): This is RIOT! (Version: 2017.04-devel-916-gd3e24a-prag-add_board_nucleo144-f746)
2017-05-04 10:47:23,536 - INFO # 
2017-05-04 10:47:23,539 - INFO # Manual UART driver test application
2017-05-04 10:47:23,542 - INFO # ===================================
2017-05-04 10:47:23,547 - INFO # This application is intended for testing additional UART
2017-05-04 10:47:23,552 - INFO # interfaces, that might be defined for a board. The 'primary' UART
2017-05-04 10:47:23,558 - INFO # interface is tested implicitly, as it is running the shell...
2017-05-04 10:47:23,558 - INFO # 
2017-05-04 10:47:23,564 - INFO # When receiving data on one of the additional UART interfaces, this
2017-05-04 10:47:23,569 - INFO # data will be outputted via STDIO. So the easiest way to test an 
2017-05-04 10:47:23,575 - INFO # UART interface, is to simply connect the RX with the TX pin. Then 
2017-05-04 10:47:23,581 - INFO # you can send data on that interface and you should see the data 
2017-05-04 10:47:23,583 - INFO # being printed to STDOUT
2017-05-04 10:47:23,583 - INFO # 
2017-05-04 10:47:23,587 - INFO # NOTE: all strings need to be '\n' terminated!
2017-05-04 10:47:23,587 - INFO # 
2017-05-04 10:47:23,588 - INFO # UART INFO:
2017-05-04 10:47:23,591 - INFO # Available devices:               3
2017-05-04 10:47:23,595 - INFO # UART used for STDIO (the shell): UART_DEV(0)
2017-05-04 10:47:23,595 - INFO # 
> init 1 9600
2017-05-04 10:47:28,817 - INFO #  init 1 9600
2017-05-04 10:47:28,820 - INFO # Successfully initialized UART_DEV(1)
> send 1 n
2017-05-04 10:47:32,642 - INFO #  send 1 n
2017-05-04 10:47:32,645 - INFO # UART_DEV(1) TX: n
2017-05-04 10:47:32,648 - INFO # UART_DEV(1) RX: n\n
> send 1 blubbbar
2017-05-04 10:47:36,467 - INFO #  send 1 blubbbar
2017-05-04 10:47:36,469 - INFO # UART_DEV(1) TX: blubbbar
> 2017-05-04 10:47:36,481 - INFO #  UART_DEV(1) RX: blubbbar\n
```

aabadie commented:
Had a look at the changes again. They all looks good to me now so it's time for squashing here.
I'll give another try next week (I won't have the board with me before tuesday) to see if the UART still doesn't work.

What's the version of your compiler ? (I have gcc 4.9.3) Maybe I can try another one to see if it makes a difference.

haukepetersen commented:
```
arm-none-eabi-gcc (GNU Tools for ARM Embedded Processors) 5.4.1 20160919 (release) [ARM/embedded-5-branch revision 240496]
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

haukepetersen commented:
squashed

haukepetersen commented:
@vincent-d: are all your change requests fulfilled?

vincent-d commented on cpu/stm32f7/vectors.c:
```diff
@@ -0,0 +1,257 @@
+    (void*) isr_rtc_wkup,
+    (void*) isr_flash,
+    (void*) isr_rcc,
+    (void*) isr_exti0,
```
It seems like every other stm32 define every `isr_exti*` as `isr_exti`, gpio interrupt are crashing otherwise.

haukepetersen replied:
Yes, that's correct. This change was lost when I copied in the correct isr table...

haukepetersen commented:
fixed `exti` ISR vectors

haukepetersen commented:
@aabadie: what is your verdict, are you ready to approve as well?

aabadie commented:
> what is your verdict, are you ready to approve as well?

I re-tested yesterday by flashing an mbed generated firmware that should print something on UART: didn't work. So there's probably something broken on the board I have.
BTW, the changes are fine here: ACK

aabadie commented:
Re-triggered a build on Murdock2

aabadie commented:
All green: Go!

soleilos commented:
Hi, For the BOARD=nucleo144-f767, the defaults and support for the I2C and SPI is not yet defined in there periph_conf.h. Is there any plan for this update?

Thank you.

haukepetersen commented:
We just need to find someone doing this, so feel free to port them :-)
